src/router.ts

```diff
@@ -0,0 +1,125 @@
+import Router from '@koa/router';
+import { z } from 'zod';
+import { fromZodError } from 'zod-validation-error';
+import zodToJsonSchema from 'zod-to-json-schema';
+import compose = require('koa-compose');
+import { IntrospectionConfig } from './koa';
+import { EndpointImplementation, implementRoute } from './koa-utils';
+import { IntrospectionResponse, OneSchemaDefinition } from './types';
+
+export type RouterEndpointDefinition<Name> = {
+  name: Name;
+  description?: string;
+  request: z.ZodType<any, any, any>;
+  response: z.ZodType<any, any, any>;
+};
+
+type Method = 'GET' | 'DELETE' | 'PUT' | 'POST' | 'PATCH';
+type RoughRoute = `${Method} ${string}`;
+
+type ZodSchema = {
+  [route: string]: RouterEndpointDefinition<string>;
+};
+
+export type OneSchemaRouterConfig<R extends Router<any, any>> = {
+  using: R;
+  introspection: IntrospectionConfig | undefined;
+};
+
+export class OneSchemaRouter<
+  Schema extends ZodSchema,
+  R extends Router<any, any>,
+> {
+  private router: R;
+
+  private constructor(
+    private schema: Schema,
+    { introspection, using: router }: OneSchemaRouterConfig<R>,
+  ) {
+    this.router = router;
+    if (introspection) {
+      router.get(introspection.route, (ctx, next) => {
+        const response: IntrospectionResponse = {
+          serviceVersion: introspection.serviceVersion,
+          schema: convertRouterSchemaToJSONSchemaStyle(this.schema),
+        };
```

Comment:
Would be nice if this followed the OpenAPI spec, so other tools could consume this schema, instead of just one-schema.

Reply:
@epeters3 I love that as a long-term goal.

Since developers shouldn't need to really interact directly with the schema anymore in the Zod-based flow, it seems reasonable that we'd stop maintaining our own custom meta-schema, and start moving towards OpenAPI.

The main things that _are_ represented in OpenAPI, and _not_ represented in the `one-schema` format are:
- responses categorized by status code (one-schema currently only cares about the successful response)
- "resources" in OpenAPI work a little differently, and could be present in one of various locations.

We'd need to find ways to account for these differences in our codegen logic before migrating to pure open-api format.

Reply:
Thanks for the response. Yes I think this is just a nice-to-have, with a non-trivial implementation burden, so waiting is fine.

Reply:
Actually, will the `generate-open-api-spec` utility work on these zod schemas? I'm assuming no, but if yes, then we wouldn't even have to worry about this.

Reply:
Yes -- that'd work as a temporary solution for someone wanting an OpenAPI spec.

```
1. Serve schema via introspection (using Zod-based solution)
2. one-schema fetch-remote-schema ...
3. one-schema generate-open-api-spec
```

Reply:
Ok cool! A quality of life improvement to that workflow could be to have the router encapsulate the `generate-open-api-spec` logic behind an `introspection/openapi` route or something. Just thinking out loud.

Reply:
@epeters3 I like it.

Reply:
Cool. I don't feel like that new endpoint is needed right now, but nice to know of a potential path forward.